calico/felix/cleanup.py

```diff
@@ -0,0 +1,100 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) Tigera Inc 2016. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+import logging
+import re
+from subprocess import check_output, check_call, call
+
+from calico.felix.frules import FELIX_PREFIX, POSTROUTING_LOCAL_NAT_FRAGMENT
+from calico.felix.ipsets import FELIX_PFX
+from calico.felix.masq import MASQ_RULE_FRAGMENT
+
+_log = logging.getLogger(__name__)
+
+JUMP_RULE_RE = r'-A ((?!%s)\w+ .*-j %s\w+.*)' % (FELIX_PREFIX, FELIX_PREFIX)
+"""Regex to match top-level jump rules from, for example, INPUT to
+felix-INPUT."""
+
+IPSET_NAME_RE = r"^Name: (%s.*)" % FELIX_PFX
+
+
+def main():
+    clean_up_iptables("iptables", "iptables-save")
+    clean_up_iptables("ip6tables", "ip6tables-save")
+    clean_up_ipsets()
+
+
+def clean_up_iptables(iptables_cmd, iptables_save_cmd):
+    # We can't delete chains until they're unreferenced.  First, remove felix
+    # jump rules from kernel chains.
+    ipt_lines = check_output([iptables_save_cmd]).splitlines()
+    table = None
+    for line in ipt_lines:
+        if line.startswith("*"):
+            # Start of a new table, save off the name.
+            table = line[1:]
```

Comment:
Think should just have line[1] here, and assert that there aren't more words on the line.  If we don't check, and there ever are extra words, line[1:] won't substitute properly where it's used in further iptables commands below.